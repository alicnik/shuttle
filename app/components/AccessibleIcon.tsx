import * as AccessibleIconComponent from '@radix-ui/react-accessible-icon';

export const AccessibleIcon = AccessibleIconComponent.Root;
