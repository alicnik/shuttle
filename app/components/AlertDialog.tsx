import {
  AlertDialog as AlertDialogComponent,
  AlertDialogAction,
  AlertDialogCancel,
  AlertDialogContent,
  AlertDialogDescription,
  AlertDialogFooter,
  AlertDialogHeader,
  AlertDialogTitle,
} from '~/components/ui/alert-dialog';

type AlertDialogProps = React.PropsWithChildren<{
  title: string;
  description: string;
  onConfirm: () => void;
}>;

export function AlertDialog({
  children,
  title,
  description,
  onConfirm,
}: AlertDialogProps) {
  return (
    <AlertDialogComponent>
      {children}
      <AlertDialogContent onCloseAutoFocus={(e) => e.preventDefault()}>
        <AlertDialogHeader>
          <AlertDialogTitle>{title}</AlertDialogTitle>
          <AlertDialogDescription>{description}</AlertDialogDescription>
        </AlertDialogHeader>
        <AlertDialogFooter>
          <AlertDialogCancel>Cancel</AlertDialogCancel>
          <AlertDialogAction onClick={onConfirm}>Continue</AlertDialogAction>
        </AlertDialogFooter>
      </AlertDialogContent>
    </AlertDialogComponent>
  );
}
