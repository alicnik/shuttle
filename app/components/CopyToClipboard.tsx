import { ClipboardCopyIcon } from '@radix-ui/react-icons';
import { useToast } from './ui';
import { Tooltip } from './Tooltip';
import type { TooltipProps } from './Tooltip';

interface CopyToClipboardProps {
  copyText: string;
  successMessage: string;
  tooltipOptions: Omit<TooltipProps, 'children'>;
}

export function CopyToClipboard({
  copyText,
  successMessage,
  tooltipOptions: { content, side = 'top', sideOffset = 0 },
}: CopyToClipboardProps) {
  const { toast } = useToast();

  return (
    <Tooltip content={content} side={side} sideOffset={sideOffset}>
      <ClipboardCopyIcon
        className="translate-y-[3px] cursor-pointer"
        onClick={async () => {
          await navigator.clipboard.writeText(copyText);
          toast({ description: successMessage });
        }}
      />
    </Tooltip>
  );
}
