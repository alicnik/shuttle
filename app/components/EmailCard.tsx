import type { Email } from '@prisma/client';
import { Checkbox } from '~/components/ui/checkbox';
import { useNavigation, useSubmit } from '@remix-run/react';
import { formatRelative } from 'date-fns';

interface EmailCardProps {
  email: Email;
  selected: string[];
  setSelected: React.Dispatch<React.SetStateAction<string[]>>;
  preview: string | null;
  setPreview: React.Dispatch<React.SetStateAction<string | null>>;
}

export function EmailCard({
  email,
  selected,
  setSelected,
  preview,
  setPreview,
}: EmailCardProps) {
  const submit = useSubmit();
  const navigation = useNavigation();

  if (navigation.state === 'submitting') {
    console.log(Object.fromEntries(navigation.formData || []));
  }

  const isOptimisticRead =
    ['submitting', 'loading'].includes(navigation.state) &&
    navigation.formData?.get('_action') === 'markRead' &&
    navigation.formData?.get('emailId') === email.id;

  const isOptimisticUnread =
    ['submitting', 'loading'].includes(navigation.state) &&
    navigation.formData?.get('_action') === 'markSelectedUnread' &&
    String(navigation.formData?.get('selected')).split(',').includes(email.id);

  const isSelected = selected.includes(email.id);
  const isPreview = preview === email.id;

  const from = email.from.startsWith('<')
    ? email.from.slice(1, -1)
    : email.from.split('<')[0].trim();

  return (
    <div
      data-selected={isSelected}
      data-in-view={isPreview}
      className="mb-4 flex gap-2 rounded-md p-2 hover:bg-zinc-800 data-[in-view=true]:bg-zinc-800 data-[selected=true]:bg-zinc-800"
    >
      <Checkbox
        className="mt-[2px]"
        checked={isSelected || isPreview}
        onCheckedChange={(checked) => {
          setSelected((selected) => {
            if (checked) {
              return [...selected, email.id];
            }

            if (isPreview) {
              setPreview(null);
            }
            return selected.filter((id) => id !== email.id);
          });
        }}
      />
      <div
        data-read={!isOptimisticUnread && (email.read || isOptimisticRead)}
        className="cursor-pointer text-left data-[read=true]:opacity-50"
        onClick={() =>
          setPreview((current) => {
            if (current) {
              submit(
                { emailId: current, _action: 'markRead' },
                { method: 'post' }
              );
            }
            return email.id;
          })
        }
      >
        <div className="mb-1 flex items-center ">
          <p className="mr-12 w-max text-sm font-bold">{from}</p>
          <p className="ml-auto w-max text-xs">
            {formatRelative(email.createdAt, new Date())}
          </p>
        </div>
        <p className="mb-1 text-sm">{email.subject}</p>
        <p className="line-clamp-2 text-xs">{email.text}</p>
      </div>
    </div>
  );
}
