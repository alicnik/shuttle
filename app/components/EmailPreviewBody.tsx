import clsx from 'clsx';
import type { Email } from '@prisma/client';
import * as React from 'react';

export function EmailPreviewBody({ html }: { html?: Email['html'] }) {
  const htmlWithBaseTag = React.useMemo(() => {
    if (!html) return '';

    // This ensures links open in a new tab and not in the iframe
    const baseTag = '<base target="_blank">';
    const headTag = '<head>';
    const headTagIndex = html.indexOf(headTag);

    if (headTagIndex === -1) {
      return html;
    }

    return (
      html.slice(0, headTagIndex + headTag.length) +
      baseTag +
      html.slice(headTagIndex + headTag.length)
    );
  }, [html]);

  return (
    <iframe
      data-testid="email-preview-body"
      className={clsx('email-preview', 'h-full w-full')}
      srcDoc={htmlWithBaseTag}
      title="Email preview"
    />
  );
}
