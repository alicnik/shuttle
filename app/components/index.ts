export * from './AlertDialog';
export * from './CopyToClipboard';
export * from './EmailCard';
export * from './EmailPreviewBody';
export * from './EmailPreviewHeader';
export * from './HarvestLinks';
export * from './Theme';
export * from './Tooltip';
export * from './ui';
