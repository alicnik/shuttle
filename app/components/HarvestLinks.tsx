import * as React from 'react';
import {
  Button,
  Dialog,
  DialogContent,
  DialogDescription,
  DialogHeader,
  DialogTitle,
  DialogTrigger,
  useToast,
} from './ui';
import { Tooltip } from './Tooltip';
import { Link2Icon } from '@radix-ui/react-icons';
import { LINK_COPY_SUCCESS_MESSAGE } from '~/lib';
import type { Email } from '@prisma/client';
import { AccessibleIcon } from './AccessibleIcon';

interface HarvestLinksProps {
  email: Email;
}

export function HarvestLinks({ email }: HarvestLinksProps) {
  const { toast } = useToast();

  const links = React.useMemo(() => {
    /* c8 ignore next 3 */
    if (typeof window === 'undefined') {
      return [];
    }
    const parser = new DOMParser();
    const doc = parser.parseFromString(email.html, 'text/html');
    const links = Array.from(doc.querySelectorAll('a'));
    const mapped = links.map((link) => {
      const linkDoc = parser.parseFromString(link.innerHTML, 'text/html');
      const elements = linkDoc.querySelectorAll('*');
      elements.forEach((element) => {
        element.removeAttribute('style');
        element.removeAttribute('width');
        element.removeAttribute('height');
        if (element.tagName === 'IMG') {
          element.setAttribute('style', 'max-width: 80%;');
        }
      });
      link.innerHTML = linkDoc.body.innerHTML;
      return link;
    });
    return mapped;
  }, [email.html]);

  return (
    <Dialog>
      <Tooltip content="Harvest links">
        <DialogTrigger asChild>
          <Button
            variant="ghost"
            className="hover:bg-zinc-300"
            type="button"
            size="icon"
          >
            <AccessibleIcon label="Harvest links from email">
              <Link2Icon />
            </AccessibleIcon>
          </Button>
        </DialogTrigger>
      </Tooltip>
      <DialogContent onCloseAutoFocus={(e) => e.preventDefault()}>
        <DialogHeader>
          <DialogTitle>Found the following links in the email</DialogTitle>
          <DialogDescription>
            Click a link to copy it to your clipboard
          </DialogDescription>
          {links.length > 0 ? (
            <ul className="grid auto-rows-fr grid-cols-2 place-items-center gap-x-2 gap-y-8 pt-8">
              {links.map((link) => {
                return (
                  <Tooltip key={link.href} content={link.href} sideOffset={4}>
                    <li
                      className="flex cursor-pointer justify-center"
                      dangerouslySetInnerHTML={{
                        __html: link.innerHTML,
                      }}
                      onClick={async () => {
                        await navigator.clipboard.writeText(link.href);
                        toast({
                          description: LINK_COPY_SUCCESS_MESSAGE,
                        });
                      }}
                    />
                  </Tooltip>
                );
              })}
            </ul>
          ) : (
            <DialogDescription>
              No links were found in the email
            </DialogDescription>
          )}
        </DialogHeader>
      </DialogContent>
    </Dialog>
  );
}
