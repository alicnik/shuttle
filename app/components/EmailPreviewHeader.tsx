import { Form, Link, useSubmit } from '@remix-run/react';
import { format } from 'date-fns';
import { Tooltip } from './Tooltip';
import { Button, Separator } from './ui';
import type { Email, User } from '@prisma/client';
import {
  EnvelopeClosedIcon,
  ExternalLinkIcon,
  TrashIcon,
} from '@radix-ui/react-icons';
import { AlertDialog } from './AlertDialog';
import { AlertDialogTrigger } from './ui/alert-dialog';

interface EmailPreviewHeaderProps {
  email: Email;
  userId: User['id'];
  setPreview: React.Dispatch<React.SetStateAction<string | null>>;
  setSelected: React.Dispatch<React.SetStateAction<string[]>>;
}

export function EmailPreviewHeader({
  email,
  userId,
  setPreview,
  setSelected,
}: EmailPreviewHeaderProps) {
  const submit = useSubmit();

  return (
    <div className="max-h-48 overflow-hidden p-4 text-zinc-900">
      <h3 className="truncate">
        <span className="font-bold">From:</span> {email.from}
      </h3>
      <h3 className="truncate">
        <span className="font-bold">To:</span> {`${email.userId}@shuttle.email`}
      </h3>
      <h3 className="truncate">
        <span className="font-bold">Subject:</span> {email.subject}
      </h3>
      <h3 className="mb-4 truncate">
        <span className="font-bold">Received:</span>{' '}
        {format(email.createdAt, 'cccc, io MMMM yyyy @ hh:mm a')}
      </h3>
      <Form method="post" className="mb-4 flex gap-2">
        <input type="hidden" name="selected" value={email.id} />
        <Tooltip content="Mark as unread">
          <Button
            variant="ghost"
            size="icon"
            name="_action"
            value="markSelectedUnread"
          >
            <EnvelopeClosedIcon />
          </Button>
        </Tooltip>
        <AlertDialog
          title="Are you sure?"
          description="This action cannot be undone."
          onConfirm={() => {
            setPreview(null);
            setSelected((selected) => selected.filter((id) => id !== email.id));
            submit(
              { selected: email.id, _action: 'deleteSelected' },
              { method: 'post' }
            );
          }}
        >
          <Tooltip content="Delete">
            <AlertDialogTrigger asChild>
              <Button
                name="_action"
                value="deleteSelected"
                variant="ghost"
                size="icon"
                className="hover:bg-destructive focus:bg-destructive"
              >
                <TrashIcon />
              </Button>
            </AlertDialogTrigger>
          </Tooltip>
        </AlertDialog>
        <Tooltip content="Open in new window">
          <Link to={`/${userId}/${email.id}`} target="_blank">
            <Button
              type="button"
              variant="ghost"
              size="icon"
              onClick={() => {
                submit(
                  { emailId: email.id, _action: 'markRead' },
                  { method: 'post' }
                );
              }}
            >
              <ExternalLinkIcon />
            </Button>
          </Link>
        </Tooltip>
      </Form>
      <Separator />
    </div>
  );
}
