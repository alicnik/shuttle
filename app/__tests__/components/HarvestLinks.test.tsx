import { screen } from '@testing-library/react';
import { setup } from '../test-utils';
import { HarvestLinks, Toaster } from '~/components';

const wrapper = ({ children }: { children: React.ReactNode }) => (
  <>
    {children}
    <Toaster />
  </>
);

const props = {
  email: {
    id: 'mock-id',
    subject: 'mock-subject',
    from: 'mock-from',
    text: 'mock-text',
    html: `
    <html>
      <head></head>
      <body>
        <a href="https://shuttle.email">Shuttle</a>
        <a href="https://mock-href.com">
          <img style="padding: 10px;" width="10" height="10" src="mock-src" />
        </a>
      </body>
    </html>
    `,
    userId: 'mock-user-id',
    read: false,
    createdAt: new Date('October 13, 2023 13:07:00'),
  },
};

describe('HarvestLinks', () => {
  it('should match the snapshot if there are links', async () => {
    const { user } = setup(<HarvestLinks {...props} />);

    await user.click(screen.getByRole('button'));

    expect(screen.getByRole('dialog')).toMatchSnapshot();
  });

  it('should match the snapshot if there are no links', async () => {
    const { user } = setup(
      <HarvestLinks email={{ ...props.email, html: '<p>No links</p>' }} />
    );

    await user.click(screen.getByRole('button'));

    expect(screen.getByRole('dialog')).toMatchSnapshot();
  });

  it('extracts the links and formats images appropriately', async () => {
    const { user } = setup(<HarvestLinks {...props} />);

    await user.click(screen.getByRole('button'));

    const image = screen.getByRole('img');

    expect(image).not.toHaveAttribute('style', 'padding: 10px;');
    expect(image).not.toHaveAttribute('width');
    expect(image).not.toHaveAttribute('height');
    expect(image).toHaveAttribute('style', 'max-width: 80%;');
  });

  it('should allow the user to click on harvested links to copy the href to the clipboard', async () => {
    const { user } = setup(<HarvestLinks {...props} />, { wrapper });

    vi.spyOn(navigator.clipboard, 'writeText');

    await user.click(screen.getByRole('button'));
    await user.click(screen.getAllByRole('listitem')[0]);

    expect(navigator.clipboard.writeText).toHaveBeenCalledWith(
      'https://shuttle.email/'
    );
    expect(screen.getByText(/link copied to clipboard/i)).toBeInTheDocument();
  });
});
