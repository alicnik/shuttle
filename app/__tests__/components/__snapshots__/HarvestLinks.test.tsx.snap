// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HarvestLinks > should match the snapshot if there are links 1`] = `
<div
  aria-describedby="radix-:r2:"
  aria-labelledby="radix-:r1:"
  class="fixed left-[50%] top-[50%] z-50 grid w-full max-w-lg translate-x-[-50%] translate-y-[-50%] gap-4 border bg-background p-6 shadow-lg duration-200 data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[state=closed]:slide-out-to-left-1/2 data-[state=closed]:slide-out-to-top-[48%] data-[state=open]:slide-in-from-left-1/2 data-[state=open]:slide-in-from-top-[48%] sm:rounded-lg md:w-full"
  data-state="open"
  id="radix-:r0:"
  role="dialog"
  style="pointer-events: auto;"
  tabindex="-1"
>
  <div
    class="flex flex-col space-y-1.5 text-center sm:text-left"
  >
    <h2
      class="text-lg font-semibold leading-none tracking-tight"
      id="radix-:r1:"
    >
      Found the following links in the email
    </h2>
    <p
      class="text-sm text-muted-foreground"
      id="radix-:r2:"
    >
      Click a link to copy it to your clipboard
    </p>
    <ul
      class="grid auto-rows-fr grid-cols-2 place-items-center gap-x-2 gap-y-8 pt-8"
    >
      <li
        class="flex cursor-pointer justify-center"
        data-state="closed"
      >
        Shuttle
      </li>
      <li
        class="flex cursor-pointer justify-center"
        data-state="closed"
      >
        <img
          src="mock-src"
          style="max-width: 80%;"
        />
        
        
      </li>
    </ul>
  </div>
  <button
    class="absolute right-4 top-4 rounded-sm opacity-70 ring-offset-background transition-opacity hover:opacity-100 focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:pointer-events-none data-[state=open]:bg-accent data-[state=open]:text-muted-foreground"
    type="button"
  >
    <svg
      class="h-4 w-4"
      fill="none"
      height="15"
      viewBox="0 0 15 15"
      width="15"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M11.7816 4.03157C12.0062 3.80702 12.0062 3.44295 11.7816 3.2184C11.5571 2.99385 11.193 2.99385 10.9685 3.2184L7.50005 6.68682L4.03164 3.2184C3.80708 2.99385 3.44301 2.99385 3.21846 3.2184C2.99391 3.44295 2.99391 3.80702 3.21846 4.03157L6.68688 7.49999L3.21846 10.9684C2.99391 11.193 2.99391 11.557 3.21846 11.7816C3.44301 12.0061 3.80708 12.0061 4.03164 11.7816L7.50005 8.31316L10.9685 11.7816C11.193 12.0061 11.5571 12.0061 11.7816 11.7816C12.0062 11.557 12.0062 11.193 11.7816 10.9684L8.31322 7.49999L11.7816 4.03157Z"
        fill="currentColor"
        fill-rule="evenodd"
      />
    </svg>
    <span
      class="sr-only"
    >
      Close
    </span>
  </button>
</div>
`;

exports[`HarvestLinks > should match the snapshot if there are no links 1`] = `
<div
  aria-describedby="radix-:r8:"
  aria-labelledby="radix-:r7:"
  class="fixed left-[50%] top-[50%] z-50 grid w-full max-w-lg translate-x-[-50%] translate-y-[-50%] gap-4 border bg-background p-6 shadow-lg duration-200 data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[state=closed]:slide-out-to-left-1/2 data-[state=closed]:slide-out-to-top-[48%] data-[state=open]:slide-in-from-left-1/2 data-[state=open]:slide-in-from-top-[48%] sm:rounded-lg md:w-full"
  data-state="open"
  id="radix-:r6:"
  role="dialog"
  style="pointer-events: auto;"
  tabindex="-1"
>
  <div
    class="flex flex-col space-y-1.5 text-center sm:text-left"
  >
    <h2
      class="text-lg font-semibold leading-none tracking-tight"
      id="radix-:r7:"
    >
      Found the following links in the email
    </h2>
    <p
      class="text-sm text-muted-foreground"
      id="radix-:r8:"
    >
      Click a link to copy it to your clipboard
    </p>
    <p
      class="text-sm text-muted-foreground"
      id="radix-:r8:"
    >
      No links were found in the email
    </p>
  </div>
  <button
    class="absolute right-4 top-4 rounded-sm opacity-70 ring-offset-background transition-opacity hover:opacity-100 focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:pointer-events-none data-[state=open]:bg-accent data-[state=open]:text-muted-foreground"
    type="button"
  >
    <svg
      class="h-4 w-4"
      fill="none"
      height="15"
      viewBox="0 0 15 15"
      width="15"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M11.7816 4.03157C12.0062 3.80702 12.0062 3.44295 11.7816 3.2184C11.5571 2.99385 11.193 2.99385 10.9685 3.2184L7.50005 6.68682L4.03164 3.2184C3.80708 2.99385 3.44301 2.99385 3.21846 3.2184C2.99391 3.44295 2.99391 3.80702 3.21846 4.03157L6.68688 7.49999L3.21846 10.9684C2.99391 11.193 2.99391 11.557 3.21846 11.7816C3.44301 12.0061 3.80708 12.0061 4.03164 11.7816L7.50005 8.31316L10.9685 11.7816C11.193 12.0061 11.5571 12.0061 11.7816 11.7816C12.0062 11.557 12.0062 11.193 11.7816 10.9684L8.31322 7.49999L11.7816 4.03157Z"
        fill="currentColor"
        fill-rule="evenodd"
      />
    </svg>
    <span
      class="sr-only"
    >
      Close
    </span>
  </button>
</div>
`;
